### PR TITLE
[PyTorch] Add grouped linear op and experimental fusion for grouped MLP

### PR DESCRIPTION
**Pull request by timmoon10**
# Description

This PR adds a grouped linear op, which can be used in the grouped MLP block in Mixture-of-Experts models. It also adds an experimental fused operation for a grouped MLP block, using a CuTe DSL kernel that computes an MXFP8 grouped GEMM and SwiGLU.

## Type of change

- [ ] Documentation change (change only to the documentation, either a fix or a new content)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Infra/Build change
- [ ] Code refactoring

## Changes

- Add a grouped linear operation
- Add a post-scaled SwiGLU op and add support for interleaving SwiGLU gate and linear units
- Add a fused operation for grouped MLP

# Checklist:

- [x] I have read and followed the [contributing guidelines](https://github.com/NVIDIA/TransformerEngine/blob/main/CONTRIBUTING.rst)
- [x] The functionality is complete
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: transformer_engine/pytorch/ops/fused/forward_grouped_mlp.py [hunk 1]
```diff
@@ -0,0 +1,453 @@
...
+
+        # Split input tensor and convert dtypes if needed
+        fc1_x = maybe_dequantize(input_, dtype)
+        fc1_xs = None
+        for quantizer in fc1_input_quantizers:
+            quantizer.set_usage(rowwise=True, columnwise=weight_requires_grad)
+            quantizer.optimize_for_gemm = True
+        fc1_xs = tex.split_quantize(fc1_x, split_sizes_cpu, fc1_input_quantizers)
+
+        # Pack data tensors
```
Comment (yaox12) [inline transformer_engine/pytorch/ops/fused/forward_grouped_mlp.py:183]: May be a silly question: are these packing and unpacking code just for verification? Or will they be in the final version?

File: transformer_engine/pytorch/ops/fused/forward_grouped_mlp.py [hunk 2]
```diff
@@ -0,0 +1,453 @@
...
+        )
+        num_non_exiting_tiles = torch.full(
+            (1,),
+            in_shape[0] // mma_tiler_mn[0],
+            dtype=torch.int,
+            device=device,
+        )
+
+        # Fused kernel for FC1 + SwiGLU + post-scale
+        fc1_kernel_out = self.grouped_gemm_swiglu_kernel()(
```
Comment (skydoorkai) [inline transformer_engine/pytorch/ops/fused/forward_grouped_mlp.py:234]: After swiglu, it usually needs to multiply with permuted_probs. Does this weighted swiglu supported?

File: transformer_engine/pytorch/ops/fused/forward_grouped_mlp.py (hunk 1 above)
Comment (timmoon10) [reply transformer_engine/pytorch/ops/fused/forward_grouped_mlp.py:183]: I'm working on getting rid of the concatenations, but the permutes are no-ops. The kernel API expects tensors with non-contiguous dims: https://github.com/NVIDIA/cudnn-frontend/blob/main/python/cudnn/grouped_gemm/grouped_gemm_swiglu/api.py#L240-L245

File: transformer_engine/pytorch/ops/fused/forward_grouped_mlp.py (hunk 2 above)
Comment (timmoon10) [reply transformer_engine/pytorch/ops/fused/forward_grouped_mlp.py:234]: Yes, the probs are passed into the kernel here: https://github.com/timmoon10/TransformerEngine/blob/46294be478f6551e2cf251283adc7529ddb2964e/transformer_engine/pytorch/ops/fused/forward_grouped_mlp.py#L264

File: transformer_engine/pytorch/ops/basic/grouped_linear.py [hunk 3]
```diff
@@ -0,0 +1,700 @@
...
+    clear_tensor_data,
+    devices_match,
+    round_up_to_nearest_multiple,
+)
+from .._common import is_quantized_tensor, maybe_dequantize
+from ..op import BasicOperation, OperationContext
+
+
+class GroupedLinear(BasicOperation):
+    """Apply multiple linear transformations: :math:``y_i = x_i W_i^T + b_i``
```
[the remark this answers is not in the document]
Comment (timmoon10) [reply transformer_engine/pytorch/ops/basic/grouped_linear.py]: The real bug is that the docstring isn't a raw string. I've updated this.

File: transformer_engine/pytorch/ops/basic/swiglu.py [hunk 4]
```diff
@@ -0,0 +1,415 @@
...
+from ..op import BasicOperation, OperationContext
+from .._common import maybe_dequantize
+
+__all__ = ["SwiGLU", "ClampedSwiGLU", "ScaledSwiGLU"]
+
+
+class SwiGLU(BasicOperation):
+    r"""Swish gated linear unit
+
+    The input tensor is split into chunks :math:``a`` and :math:``b``
```
[the remark this answers is not in the document]
Comment (timmoon10) [reply transformer_engine/pytorch/ops/basic/swiglu.py:26]: Double backticks are correct, and they're used elsewhere in the codebase. I'm not certain why, but there's something in the combination of Sphinx and r-strings that causes parsing errors with single backticks.

File: transformer_engine/pytorch/ops/basic/swiglu.py [hunk 5]
```diff
@@ -0,0 +1,415 @@
...
+    ) -> tuple[
+        torch.Tensor,
+        Iterable[Iterable[Optional[torch.Tensor]]],
+        Iterable[Iterable[Optional[torch.Tensor]]],
+    ]:
+        ctx = basic_op_ctxs[0]
+        input_, scales = ctx.saved_tensors
+        input_ = maybe_dequantize(input_, ctx.dtype)
+        if scales is not None:
+            scales = maybe_dequantize(scales, ctx.dtype)
```
[the remark this answers is not in the document]
Comment (timmoon10) [reply transformer_engine/pytorch/ops/basic/swiglu.py:456]: [`torch.linalg.vecdot`](https://docs.pytorch.org/docs/stable/generated/torch.linalg.vecdot.html) computes the dot product over the last tensor dim and leave all other dims unchanged. The existing code is correct and the suggested code exposes an unnecessary memory-bound kernel.

File: transformer_engine/pytorch/ops/basic/swiglu.py [hunk 6]
```diff
@@ -0,0 +1,415 @@
...
+                )
+                grad_input = grad_input.transpose(1, 2).contiguous()
+                grad_input = grad_input.view(shape)
+
+        # Compute scales grad by recomputing SwiGLU
+        grad_extra_input = None
+        if ctx.extra_input_requires_grad:
+            swiglu_out = tex.swiglu(swiglu_in, None)
+            grad_extra_input = torch.linalg.vecdot(swiglu_out, grad_output)
+
```
[the remark this answers is not in the document]
Comment (timmoon10) [reply transformer_engine/pytorch/ops/basic/swiglu.py:494]: This suggestion is incorrect, [`torch.linalg.vecdot`](https://docs.pytorch.org/docs/stable/generated/torch.linalg.vecdot.html) only reduces over the last dim.

File: transformer_engine/pytorch/ops/basic/swiglu.py [hunk 7]
```diff
@@ -0,0 +1,415 @@
...
+        # Compute scaled SwiGLU
+        swiglu_out = tex.swiglu(swiglu_in, None)
+        out = swiglu_out * scales.unsqueeze(-1)
+
+        # Save state for backward pass
+        ctx = basic_op_ctxs[0]
+        if ctx.requires_grad:
+            if is_cpu_offload_enabled():
+                mark_activation_offload(input_)
+            ctx.input_requires_grad = True
```
[the remark this answers is not in the document]
Comment (timmoon10) [reply transformer_engine/pytorch/ops/basic/swiglu.py:431]: `input_` may be a `QuantizedTensorStorage`, which does not keep track of `requires_grad`. Assuming `input_requires_grad=True` is needed for safety. This option is here for future-proofing, in case `QuantizedTensorStorage` ever gains support for tracking grads.

File: transformer_engine/pytorch/ops/basic/swiglu.py [hunk 8]
```diff
@@ -0,0 +1,415 @@
...
+            ctx.extra_input_requires_grad = extra_input.requires_grad
+            ctx.dtype = dtype
+            ctx.save_for_backward(
+                input_,
+                scales if ctx.input_requires_grad else None,
```
[the remark this answers is not in the document]
Comment (timmoon10) [reply transformer_engine/pytorch/ops/basic/swiglu.py:436]: `input_` may be a `QuantizedTensorStorage`, which does not keep track of `requires_grad`. Assuming `input_requires_grad=True` is needed for safety. This option is here for future-proofing, in case `QuantizedTensorStorage` ever gains support for tracking grads.

File: transformer_engine/pytorch/ops/fused/backward_grouped_mlp.py [hunk 9]
```diff
@@ -0,0 +1,484 @@
...
+            or window[2].out_features % 256 != 0
+        ):
+            matches_pattern = False
+        elif window[1].glu_interleave_size != 32:
+            matches_pattern = False
+
+        if matches_pattern:
+            # Construct fused op if window matches pattern
+            op = BackwardGroupedMLP_CuTeGEMMDSwiGLU_MXFP8(
+                fc1=window[0],
```
[the remark this answers is not in the document]
Comment (timmoon10) [reply transformer_engine/pytorch/ops/fused/backward_grouped_mlp.py:460]: I think your reading of the code is incorrect, although I would take suggestions for improving clarity. The loop condition guarantees the window size is 3 at the beginning of each iteration. We either fuse the window or eject the first op in the window, and then we refill back up to window size 3.

The only hint of an edge case I can see is if we perform a fusion, and that fused op can participate in further fusions. Then we might want to rewind the sliding window so that we reexamine the fused op in each window position. However, we know that the fused op is final, and we can safely advance the window past it.

**Comment by ksivaman** [comment]
Duplicates #2769
